### ISSUE TITLE
Shadowing error with custom snapshot in latest git master, not in 0.1.4

### ISSUE THREAD

**Issue by hesselink** [commercialhaskell/stack#944]
When I try to install HDBC-postgresql in a custom snapshot, it fails with the latest git master (Version 0.1.4.1, Git revision eab3db27bd27d22a2a5b4430aff7038d891d5d51 X86_64), but not with 0.1.4. The error is:

```
    <command line>: cannot satisfy -package-id parsec-3.1.9-ee2d6f78f4a115961f59c07ee210347a:
        parsec-3.1.9-ee2d6f78f4a115961f59c07ee210347a is shadowed by package parsec-3.1.9-26e7a52d0873abf27275d79645acd443
        (use -v for more information)
```

The command that failed is:

```
      /usr/bin/runhaskell -clear-package-db -global-package-db -package-db=/Users/erik/.stack/snapshots/x86_64-osx/custom-temp-snapshot/7.8.3/pkgdb/ -package-db=/Users/erik/temp/.stack-work/install/x86_64-osx/custom-temp-snapshot/7.8.3/pkgdb/ -hide-all-packages -package=Cabal-1.18.1.3 -package-id=HDBC-2.4.0.1-cb51de36851bbf1f4a1a721dc39f33f6 -package-id=array-0.5.0.0-98aa445e59f3eb0c886795ff07406d84 -package-id=base-4.7.0.1-c64d224738ec7af4085e89ca9c12c37b -package-id=bin-package-db-0.0.0.0-d82393a6f8b94918b5cee098483a704c -package-id=binary-0.7.1.0-06fe2cfe2c00e6f7d43cdb53345c4945 -package-id=bytestring-0.10.4.0-18fe2f3ce284617c82da1702e16772cf -package-id=containers-0.5.5.1-23e2a2b94d6e452c773209f31d8672c5 -package-id=convertible-1.1.1.0-83d303e51209afe187c2667409bbee0d -package-id=deepseq-1.3.0.2-733fe43e121f761739636bd6670bea77 -package-id=directory-1.2.1.0-bfdf4eb032907ecaf13c26632da610b8 -package-id=filepath-1.3.0.2-1580a61d3226e4be45fe2130dc2881e3 -package-id=ghc-prim-0.3.1.0-954cb57749cf319beafdc89b3415422c -package-id=haskeline-0.7.1.2-ec389140730241a5428bc5563724c813 -package-id=hoopl-3.10.0.1-78e45fe2aae96846315ab833b31e7409 -package-id=hpc-0.6.0.1-fa99203f9c565a963745ce11860c456f -package-id=integer-gmp-0.5.1.0-d42e6a7874a019e6a0d1c7305ebc83c4 -package-id=mtl-2.1.3.1-8bcc0591131896cfc8761a93703d4c61 -package-id=mtl-2.2.1-05c98059bc3926647251ab9e6cb3164e -package-id=old-locale-1.0.0.6-09baf1dbc5be8338e5eba7c5bb515505 -package-id=old-time-1.1.0.2-eab97c1e6d2cda84575b30ce4bfcfaf2 -package-id=parsec-3.1.9-26e7a52d0873abf27275d79645acd443 -package-id=parsec-3.1.9-ee2d6f78f4a115961f59c07ee210347a -package-id=pretty-1.1.1.1-055fb3b6b88c48ad1bf91e5cf35828d1 -package-id=process-1.2.0.0-4870cdb503fdbb301e26c7b926e216b2 -package-id=setenv-0.1.1.3-9a7f4ba08f42b64c70f174e426f2de6d -package-id=tar-0.4.1.0-849d8b66d009401931a74b583afcfa06 -package-id=template-haskell-2.9.0.0-310613e57f38a482326cc4ba2989009e -package-id=terminfo-0.4.0.0-b02cf5922e12c1a9557df1270cdf18b7 -package-id=text-1.2.0.6-e9cc2f9cbc7fea487488c8ed44ced4c0 -package-id=time-1.4.2-bf925e935c287d0b75398fe297453c28 -package-id=transformers-0.3.0.0-16a97696ae672940f1523b262e566ba5 -package-id=unix-2.7.0.1-299ecd8d04080b4c760a1a3bcfad059b -package-id=utf8-string-1.0.1.1-ed6a73bde3175c612b542512a65fc45e -package-id=xhtml-3000.2.1-2da8c2cd0581a1aa5b8f1c06f5c3b943 -package-id=zlib-0.5.4.2-4ebdf7bad9192e1149e8553ecd1c3649 /var/folders/qg/4lbbbp0d1hbg7f6_cr2r8h8h0000gn/T/stack13539/HDBC-postgresql-2.3.2.3/Setup.hs --builddir=.stack-work/dist/x86_64-osx/Cabal-1.18.1.3/ configure --user --package-db=clear --package-db=global --package-db=/Users/erik/.stack/snapshots/x86_64-osx/custom-temp-snapshot/7.8.3/pkgdb/ --libdir=/Users/erik/.stack/snapshots/x86_64-osx/custom-temp-snapshot/7.8.3/lib --bindir=/Users/erik/.stack/snapshots/x86_64-osx/custom-temp-snapshot/7.8.3/bin --datadir=/Users/erik/.stack/snapshots/x86_64-osx/custom-temp-snapshot/7.8.3/share --libexecdir=/Users/erik/.stack/snapshots/x86_64-osx/custom-temp-snapshot/7.8.3/libexec --sysconfdir=/Users/erik/.stack/snapshots/x86_64-osx/custom-temp-snapshot/7.8.3/etc --docdir=/Users/erik/.stack/snapshots/x86_64-osx/custom-temp-snapshot/7.8.3/doc/HDBC-postgresql-2.3.2.3 --htmldir=/Users/erik/.stack/snapshots/x86_64-osx/custom-temp-snapshot/7.8.3/doc/HDBC-postgresql-2.3.2.3 --haddockdir=/Users/erik/.stack/snapshots/x86_64-osx/custom-temp-snapshot/7.8.3/doc/HDBC-postgresql-2.3.2.3 --constraint=HDBC==2.4.0.1 --constraint=base==4.7.0.1 --constraint=bytestring==0.10.4.0 --constraint=convertible==1.1.1.0 --constraint=mtl==2.2.1 --constraint=old-locale==1.0.0.6 --constraint=old-time==1.1.0.2 --constraint=parsec==3.1.9 --constraint=time==1.4.2 --constraint=utf8-string==1.0.1.1 -f-mintime15
```

The corresponding command from the succeeding run with 0.1.4 is:

```
2015-09-07 16:52:39.064039: [debug] Run process: /usr/bin/runhaskell -package=Cabal-1.18.1.3 -clear-package-db -global-package-db -package-db=/Users/erik/.stack/snapshots/x86_64-osx/custom-temp-snapshot/7.8.3/pkgdb/ /var/folders/qg/4lbbbp0d1hbg7f6_cr2r8h8h0000gn/T/stack15424/HDBC-postgresql-2.3.2.3/Setup.hs --builddir=.stack-work/dist/x86_64-osx/Cabal-1.18.1.3/ configure --user --package-db=clear --package-db=global --package-db=/Users/erik/.stack/snapshots/x86_64-osx/custom-temp-snapshot/7.8.3/pkgdb/ --libdir=/Users/erik/.stack/snapshots/x86_64-osx/custom-temp-snapshot/7.8.3/lib --bindir=/Users/erik/.stack/snapshots/x86_64-osx/custom-temp-snapshot/7.8.3/bin --datadir=/Users/erik/.stack/snapshots/x86_64-osx/custom-temp-snapshot/7.8.3/share --libexecdir=/Users/erik/.stack/snapshots/x86_64-osx/custom-temp-snapshot/7.8.3/libexec --sysconfdir=/Users/erik/.stack/snapshots/x86_64-osx/custom-temp-snapshot/7.8.3/etc --docdir=/Users/erik/.stack/snapshots/x86_64-osx/custom-temp-snapshot/7.8.3/doc/HDBC-postgresql-2.3.2.3 --htmldir=/Users/erik/.stack/snapshots/x86_64-osx/custom-temp-snapshot/7.8.3/doc/HDBC-postgresql-2.3.2.3 --haddockdir=/Users/erik/.stack/snapshots/x86_64-osx/custom-temp-snapshot/7.8.3/doc/HDBC-postgresql-2.3.2.3 --constraint=HDBC==2.4.0.1 --constraint=base==4.7.0.1 --constraint=bytestring==0.10.4.0 --constraint=convertible==1.1.1.0 --constraint=mtl==2.2.1 --constraint=old-locale==1.0.0.6 --constraint=old-time==1.1.0.2 --constraint=parsec==3.1.9 --constraint=time==1.4.2 --constraint=utf8-string==1.0.1.1 -f-mintime15 @(stack_FhlmuwqZDji0WhaDijVRMV:Stack.Build.Execute src/Stack/Build/Execute.hs:696:21)
```

These are the packages it was installing:

```
Would build:
HDBC-2.4.0.1: database=snapshot, source=package index, after: convertible-1.1.1.0,mtl-2.2.1,text-1.2.1.3,utf8-string-1.0.1.1
HDBC-postgresql-2.3.2.3: database=snapshot, source=package index, after: HDBC-2.4.0.1,convertible-1.1.1.0,mtl-2.2.1,parsec-3.1.9,utf8-string-1.0.1.1
convertible-1.1.1.0: database=snapshot, source=package index, after: mtl-2.2.1,text-1.2.1.3
mtl-2.2.1: database=snapshot, source=package index, after: transformers-0.4.3.0
parsec-3.1.9: database=snapshot, source=package index, after: mtl-2.2.1,text-1.2.1.3
text-1.2.1.3: database=snapshot, source=package index
transformers-0.4.3.0: database=snapshot, source=package index
utf8-string-1.0.1.1: database=snapshot, source=package index
```

The output from `ghc-pkg`:

```
$ stack exec ghc-pkg list parsec
/Library/Frameworks/GHC.framework/Versions/7.8.3-x86_64/usr/lib/ghc-7.8.3/package.conf.d
   parsec-3.1.9
/Users/erik/.stack/snapshots/x86_64-osx/custom-temp-snapshot/7.8.3/pkgdb
   parsec-3.1.9
/Users/erik/temp/.stack-work/install/x86_64-osx/custom-temp-snapshot/7.8.3/pkgdb
```

I can post the full snapshot and stack.yaml, as well as the full build output (with -v) from both commands if that's useful.

**Comment by snoyberg** [comment]
Can you run `stack exec -- ghc-pkg list parsec -v`, to provide the installed package IDs in the output?

I'm not sure what would have changed in stack to trigger this, but I'm not surprised by this behavior. There are limitations to Cabal 1.18 (which ships with 7.8) regarding specifying packages by installed ID. Installing a newer Cabal library- while slightly dangerous- may fix this. You'd need the Cabal library to end up in the global package database.

**Comment by hesselink** [comment]
```
$ stack exec -- ghc-pkg list parsec -v
Timestamp 2015-09-07 16:34:31 UTC for /Users/erik/temp/.stack-work/install/x86_64-osx/custom-temp-snapshot/7.8.3/pkgdb/package.cache
Timestamp 2015-09-07 16:34:31 UTC for /Users/erik/temp/.stack-work/install/x86_64-osx/custom-temp-snapshot/7.8.3/pkgdb (same as cache)
using cache: /Users/erik/temp/.stack-work/install/x86_64-osx/custom-temp-snapshot/7.8.3/pkgdb/package.cache
Timestamp 2015-09-07 16:52:46 UTC for /Users/erik/.stack/snapshots/x86_64-osx/custom-temp-snapshot/7.8.3/pkgdb/package.cache
Timestamp 2015-09-07 16:52:46 UTC for /Users/erik/.stack/snapshots/x86_64-osx/custom-temp-snapshot/7.8.3/pkgdb (same as cache)
using cache: /Users/erik/.stack/snapshots/x86_64-osx/custom-temp-snapshot/7.8.3/pkgdb/package.cache
Timestamp 2015-05-29 13:24:14 UTC for /Library/Frameworks/GHC.framework/Versions/7.8.3-x86_64/usr/lib/ghc-7.8.3/package.conf.d/package.cache
Timestamp 2015-05-29 13:24:14 UTC for /Library/Frameworks/GHC.framework/Versions/7.8.3-x86_64/usr/lib/ghc-7.8.3/package.conf.d (same as cache)
using cache: /Library/Frameworks/GHC.framework/Versions/7.8.3-x86_64/usr/lib/ghc-7.8.3/package.conf.d/package.cache
/Library/Frameworks/GHC.framework/Versions/7.8.3-x86_64/usr/lib/ghc-7.8.3/package.conf.d
   parsec-3.1.9 (parsec-3.1.9-26e7a52d0873abf27275d79645acd443)
/Users/erik/.stack/snapshots/x86_64-osx/custom-temp-snapshot/7.8.3/pkgdb
   parsec-3.1.9 (parsec-3.1.9-ee2d6f78f4a115961f59c07ee210347a)
/Users/erik/temp/.stack-work/install/x86_64-osx/custom-temp-snapshot/7.8.3/pkgdb
```

I installed Cabal 1.22.4.0 globally, wiped the custom snapshot and tried again. I got the same error, although I can see it's now using the new Cabal library:

```
--  While building package HDBC-postgresql-2.3.2.3 using:
      /usr/bin/runhaskell -clear-package-db -global-package-db -package-db=/Users/erik/.stack/snapshots/x86_64-osx/custom-temp-snapshot/7.8.3/pkgdb/ -package-db=/Users/erik/temp/.stack-work/install/x86_64-osx/custom-temp-snapshot/7.8.3/pkgdb/ -hide-all-packages -package=Cabal-1.22.4.0 -package-id=HDBC-2.4.0.1-cb51de36851bbf1f4a1a721dc39f33f6 -package-id=array-0.5.0.0-98aa445e59f3eb0c886795ff07406d84 -package-id=base-4.7.0.1-c64d224738ec7af4085e89ca9c12c37b -package-id=bin-package-db-0.0.0.0-d82393a6f8b94918b5cee098483a704c -package-id=binary-0.7.1.0-06fe2cfe2c00e6f7d43cdb53345c4945 -package-id=bytestring-0.10.4.0-18fe2f3ce284617c82da1702e16772cf -package-id=containers-0.5.5.1-23e2a2b94d6e452c773209f31d8672c5 -package-id=convertible-1.1.1.0-83d303e51209afe187c2667409bbee0d -package-id=deepseq-1.3.0.2-733fe43e121f761739636bd6670bea77 -package-id=directory-1.2.1.0-bfdf4eb032907ecaf13c26632da610b8 -package-id=filepath-1.3.0.2-1580a61d3226e4be45fe2130dc2881e3 -package-id=ghc-prim-0.3.1.0-954cb57749cf319beafdc89b3415422c -package-id=haskeline-0.7.1.2-ec389140730241a5428bc5563724c813 -package-id=hoopl-3.10.0.1-78e45fe2aae96846315ab833b31e7409 -package-id=hpc-0.6.0.1-fa99203f9c565a963745ce11860c456f -package-id=integer-gmp-0.5.1.0-d42e6a7874a019e6a0d1c7305ebc83c4 -package-id=mtl-2.1.3.1-8bcc0591131896cfc8761a93703d4c61 -package-id=mtl-2.2.1-05c98059bc3926647251ab9e6cb3164e -package-id=old-locale-1.0.0.6-09baf1dbc5be8338e5eba7c5bb515505 -package-id=old-time-1.1.0.2-eab97c1e6d2cda84575b30ce4bfcfaf2 -package-id=parsec-3.1.9-26e7a52d0873abf27275d79645acd443 -package-id=parsec-3.1.9-ee2d6f78f4a115961f59c07ee210347a -package-id=pretty-1.1.1.1-055fb3b6b88c48ad1bf91e5cf35828d1 -package-id=process-1.2.0.0-4870cdb503fdbb301e26c7b926e216b2 -package-id=setenv-0.1.1.3-9a7f4ba08f42b64c70f174e426f2de6d -package-id=tar-0.4.1.0-849d8b66d009401931a74b583afcfa06 -package-id=template-haskell-2.9.0.0-310613e57f38a482326cc4ba2989009e -package-id=terminfo-0.4.0.0-b02cf5922e12c1a9557df1270cdf18b7 -package-id=text-1.2.0.6-e9cc2f9cbc7fea487488c8ed44ced4c0 -package-id=time-1.4.2-bf925e935c287d0b75398fe297453c28 -package-id=transformers-0.3.0.0-16a97696ae672940f1523b262e566ba5 -package-id=unix-2.7.0.1-299ecd8d04080b4c760a1a3bcfad059b -package-id=utf8-string-1.0.1.1-ed6a73bde3175c612b542512a65fc45e -package-id=xhtml-3000.2.1-2da8c2cd0581a1aa5b8f1c06f5c3b943 -package-id=zlib-0.5.4.2-4ebdf7bad9192e1149e8553ecd1c3649 /var/folders/qg/4lbbbp0d1hbg7f6_cr2r8h8h0000gn/T/stack73225/HDBC-postgresql-2.3.2.3/Setup.hs --builddir=.stack-work/dist/x86_64-osx/Cabal-1.22.4.0/ configure --user --package-db=clear --package-db=global --package-db=/Users/erik/.stack/snapshots/x86_64-osx/custom-temp-snapshot/7.8.3/pkgdb/ --libdir=/Users/erik/.stack/snapshots/x86_64-osx/custom-temp-snapshot/7.8.3/lib --bindir=/Users/erik/.stack/snapshots/x86_64-osx/custom-temp-snapshot/7.8.3/bin --datadir=/Users/erik/.stack/snapshots/x86_64-osx/custom-temp-snapshot/7.8.3/share --libexecdir=/Users/erik/.stack/snapshots/x86_64-osx/custom-temp-snapshot/7.8.3/libexec --sysconfdir=/Users/erik/.stack/snapshots/x86_64-osx/custom-temp-snapshot/7.8.3/etc --docdir=/Users/erik/.stack/snapshots/x86_64-osx/custom-temp-snapshot/7.8.3/doc/HDBC-postgresql-2.3.2.3 --htmldir=/Users/erik/.stack/snapshots/x86_64-osx/custom-temp-snapshot/7.8.3/doc/HDBC-postgresql-2.3.2.3 --haddockdir=/Users/erik/.stack/snapshots/x86_64-osx/custom-temp-snapshot/7.8.3/doc/HDBC-postgresql-2.3.2.3 --dependency=HDBC=HDBC-2.4.0.1-cb51de36851bbf1f4a1a721dc39f33f6 --dependency=base=base-4.7.0.1-c64d224738ec7af4085e89ca9c12c37b --dependency=bytestring=bytestring-0.10.4.0-18fe2f3ce284617c82da1702e16772cf --dependency=convertible=convertible-1.1.1.0-83d303e51209afe187c2667409bbee0d --dependency=mtl=mtl-2.2.1-05c98059bc3926647251ab9e6cb3164e --dependency=old-locale=old-locale-1.0.0.6-09baf1dbc5be8338e5eba7c5bb515505 --dependency=old-time=old-time-1.1.0.2-eab97c1e6d2cda84575b30ce4bfcfaf2 --dependency=parsec=parsec-3.1.9-ee2d6f78f4a115961f59c07ee210347a --dependency=time=time-1.4.2-bf925e935c287d0b75398fe297453c28 --dependency=utf8-string=utf8-string-1.0.1.1-ed6a73bde3175c612b542512a65fc45e -f-mintime15
    Process exited with code: ExitFailure 1
    Logs have been written to: /Users/erik/temp/.stack-work/logs/HDBC-postgresql-2.3.2.3.log

    <command line>: cannot satisfy -package-id parsec-3.1.9-ee2d6f78f4a115961f59c07ee210347a:
        parsec-3.1.9-ee2d6f78f4a115961f59c07ee210347a is shadowed by package parsec-3.1.9-26e7a52d0873abf27275d79645acd443
        (use -v for more information)
```

I noticed that the git version of stack passes package ids explicitly where the older one doesn't. For parsec, it passes *both* installed parsec versions. Could that be the problem, or is that supposed to be possible?

**Comment by snoyberg** [comment]
OK, I misunderstood where the problem was coming from originally. This is caused by the new Setup.hs logic (so pinging @mitchellwrosen). I'm not completely certain though what GHC is complaining about, still investigating.

**Comment by snoyberg** [comment]
OK, got it: it's definitely a problem with stack, and I think it's coming from the work I did with global databases. If you look closely, we're passing *three* -package-id flags to GHC around parsec. We just need to make sure we pass one, with the correct ID. I'll make a tweak now, and ping you to please test it when done.

**Comment by snoyberg** [comment]
Alright, I just pushed a fix for this, I think it should do the job. Details are pretty straightforward: instead of passing around a `Set GhcPkgId`, which has no way to dedupe two different installs of the same package, we now pass around a `Map PackageIdentifier GhcPkgId`. And as an added bonus, before passing to runghc, I convert that to a `Map PackageName GhcPkgId`, so that only one copy of a package is used at any time.

**Comment by hesselink** [comment]
Yes, that seems to have fixed it, thanks! I've tried with both Cabal 1.22 and 1.18, and both work.
